zipline-loader/src/jniMain/kotlin/app/cash/zipline/loader/internal/EcdsaP256.kt

```diff
@@ -0,0 +1,136 @@
+import java.security.KeyFactory
+import java.security.KeyPairGenerator
+import java.security.PrivateKey
+import java.security.PublicKey
+import java.security.SecureRandom
+import java.security.Signature
+import java.security.interfaces.ECPublicKey
+import java.security.spec.ECGenParameterSpec
+import java.security.spec.ECParameterSpec
+import java.security.spec.ECPoint
+import java.security.spec.ECPublicKeySpec
+import java.security.spec.PKCS8EncodedKeySpec
+import okio.Buffer
+import okio.ByteString
+import okio.ByteString.Companion.toByteString
+
+/**
+ * ECDSA using the P-256 (secp256r1) curve.
+ *
+ * We've got (at least) two choices for encoding public keys as bytes:
+ *
+ *  * Java's format. This is the ASN.1 of the `SubjectPublicKeyInfo` from X.509, encoded with the
+ *    `X509EncodedKeySpec` class.
+ *  * Apple's format. This is ANSI X9.62, encoded as 3 fixed-width values concatenated together.
+ *
+ * We use Apple's format because it's more compact, and because it's easy enough to support on both
+ * platforms.
+ *
+ * Private keys are encoded using PKCS8.
+ *
+ * https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_as_data
+ */
+class EcdsaP256(
+  private val random: SecureRandom,
+) : SignatureAlgorithm {
+
+  /** Generate a key pair using the P-256 curve. */
+  fun generateP256KeyPair(): KeyPair {
+    val keyPairGenerator = KeyPairGenerator.getInstance("EC")
+    keyPairGenerator.initialize(ECGenParameterSpec("secp256r1"), random)
+    val keyPair = keyPairGenerator.generateKeyPair()
+    return KeyPair(
+      publicKey = (keyPair.public as ECPublicKey).encodeAnsiX963(),
+      privateKey = keyPair.private.encoded.toByteString(),
+    )
+  }
+
+  fun sign(privateKey: PrivateKey, plaintext: ByteString): ByteString {
+    val signer = Signature.getInstance("SHA256withECDSA")
+    signer.initSign(privateKey, random)
+    signer.update(plaintext.toByteArray())
+    return signer.sign().toByteString()
+  }
+
+  override fun sign(message: ByteString, privateKey: ByteString): ByteString {
+    return sign(decodePkcs8EcPrivateKey(privateKey), message)
+  }
+
+  fun verify(publicKey: PublicKey, signature: ByteString, data: ByteString): Boolean {
+    val verifier = Signature.getInstance("SHA256withECDSA")
+    verifier.initVerify(publicKey)
+    verifier.update(data.toByteArray())
+    return verifier.verify(signature.toByteArray())
+  }
+
+  override fun verify(message: ByteString, signature: ByteString, publicKey: ByteString): Boolean {
+    return verify(publicKey.decodeAnsiX963(), signature, message)
+  }
+}
+
+internal fun decodePkcs8EcPrivateKey(privateKey: ByteString): PrivateKey {
+  val keyFactory = KeyFactory.getInstance("EC")
+  val keySpec = PKCS8EncodedKeySpec(privateKey.toByteArray())
+  return keyFactory.generatePrivate(keySpec)
+}
+
+/** ANSI X9.63 standard concatenates `0x04 || X || Y`. */
+internal fun ECPublicKey.encodeAnsiX963(): ByteString {
```

Comment:
These are new! And they need isolated tests. Right now it’s just high-level tests.